route-rs-runtime/src/link/sync_link.rs

```diff
@@ -1,22 +1,73 @@
 use crate::element::Element;
-use crate::link::PacketStream;
+use crate::link::{ElementLinkBuilder, Link, LinkBuilder, PacketStream};
 use futures::{Async, Poll, Stream};
 
-pub struct SyncLink<E: Element> {
-    input_stream: PacketStream<E::Input>,
+#[derive(Default)]
+pub struct SyncLinkBuilder<E: Element> {
+    in_streams: Option<Vec<PacketStream<E::Input>>>,
+    element: Option<E>,
+}
+
+impl<E: Element> SyncLinkBuilder<E> {
+    pub fn new() -> Self {
+        SyncLinkBuilder {
+            in_streams: None,
+            element: None,
+        }
+    }
+}
+
+/// Although `Link` allows an arbitrary number of ingressors and egressors, `SyncLink`
+/// may only have one ingress and egress stream since it lacks some kind of queue
+/// storage. In the future we might decide to restrict the interface for this link
+/// for clearer intent.
+impl<E: Element + Send + 'static> LinkBuilder<E::Input, E::Output> for SyncLinkBuilder<E> {
+    fn ingressors(self, in_streams: Vec<PacketStream<E::Input>>) -> Self {
+        assert_eq!(in_streams.len(), 1, "SyncLink may only take 1 input stream");
+
+        SyncLinkBuilder {
+            in_streams: Some(in_streams),
+            element: self.element,
```

Comment:
I think this pattern of builder methods passing all of `self`'s state except for the input is going to become common. I'm not sure if anything can be done to abstract that idea, but we should wait until every `Link` is migrated before trying.